### ISSUE TITLE
Failed to build gem native extension on Lion 10.7.1

### ISSUE THREAD

**Issue by thibaudgg** [johnbintz/jasmine-headless-webkit#58]
Hi John,

I have an issue when I try to install jasmine-headless-webkit on my MacBoook (Core 2 Duo, Lion 10.7.1), but I successfully installed it yesterday on my iMac (i7, Lion 10.7.1) and I have the same setup on both:
- ruby 1.9.2p290 (2011-07-09 revision 32553) [x86_64-darwin11.0.0]
- XCode from the App Store
- Qt version 4.7.3 (installed with `brew install qt`)

I got: 

```
Building native extensions.  This could take a while...
ERROR:  Error installing jasmine-headless-webkit:
    ERROR: Failed to build gem native extension.

        /Users/Thibaud/.rvm/rubies/ruby-1.9.2-p290/bin/ruby extconf.rb
g++ -c -pipe -O2 -Wall -W -DQT_NO_DEBUG -DQT_WEBKIT_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/local/Cellar/qt/4.7.3/mkspecs/macx-g++ -I. -I/usr/local/Cellar/qt/4.7.3/lib/QtCore.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtCore -I/usr/local/Cellar/qt/4.7.3/lib/QtNetwork.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtNetwork -I/usr/local/Cellar/qt/4.7.3/lib/QtGui.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtGui -I/usr/local/Cellar/qt/4.7.3/lib/QtWebKit.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtWebKit -I/usr/local/Cellar/qt/4.7.3/include -I. -F/usr/local/Cellar/qt/4.7.3/lib -o Page.o HeadlessSpecRunner/Page.cpp
HeadlessSpecRunner/Page.cpp:13: warning: unused parameter ‘frame’
HeadlessSpecRunner/Page.cpp:13: warning: unused parameter ‘msg’
HeadlessSpecRunner/Page.cpp:23: warning: unused parameter ‘frame’
g++ -c -pipe -O2 -Wall -W -DQT_NO_DEBUG -DQT_WEBKIT_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/local/Cellar/qt/4.7.3/mkspecs/macx-g++ -I. -I/usr/local/Cellar/qt/4.7.3/lib/QtCore.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtCore -I/usr/local/Cellar/qt/4.7.3/lib/QtNetwork.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtNetwork -I/usr/local/Cellar/qt/4.7.3/lib/QtGui.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtGui -I/usr/local/Cellar/qt/4.7.3/lib/QtWebKit.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtWebKit -I/usr/local/Cellar/qt/4.7.3/include -I. -F/usr/local/Cellar/qt/4.7.3/lib -o Runner.o HeadlessSpecRunner/Runner.cpp
g++ -c -pipe -O2 -Wall -W -DQT_NO_DEBUG -DQT_WEBKIT_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/local/Cellar/qt/4.7.3/mkspecs/macx-g++ -I. -I/usr/local/Cellar/qt/4.7.3/lib/QtCore.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtCore -I/usr/local/Cellar/qt/4.7.3/lib/QtNetwork.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtNetwork -I/usr/local/Cellar/qt/4.7.3/lib/QtGui.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtGui -I/usr/local/Cellar/qt/4.7.3/lib/QtWebKit.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtWebKit -I/usr/local/Cellar/qt/4.7.3/include -I. -F/usr/local/Cellar/qt/4.7.3/lib -o specrunner.o specrunner.cpp
/usr/local/Cellar/qt/4.7.3/bin/moc -DQT_NO_DEBUG -DQT_WEBKIT_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/local/Cellar/qt/4.7.3/mkspecs/macx-g++ -I. -I/usr/local/Cellar/qt/4.7.3/lib/QtCore.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtCore -I/usr/local/Cellar/qt/4.7.3/lib/QtNetwork.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtNetwork -I/usr/local/Cellar/qt/4.7.3/lib/QtGui.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtGui -I/usr/local/Cellar/qt/4.7.3/lib/QtWebKit.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtWebKit -I/usr/local/Cellar/qt/4.7.3/include -I. -F/usr/local/Cellar/qt/4.7.3/lib -D__APPLE__ -D__GNUC__ HeadlessSpecRunner/Page.h -o moc_Page.cpp
g++ -c -pipe -O2 -Wall -W -DQT_NO_DEBUG -DQT_WEBKIT_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/local/Cellar/qt/4.7.3/mkspecs/macx-g++ -I. -I/usr/local/Cellar/qt/4.7.3/lib/QtCore.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtCore -I/usr/local/Cellar/qt/4.7.3/lib/QtNetwork.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtNetwork -I/usr/local/Cellar/qt/4.7.3/lib/QtGui.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtGui -I/usr/local/Cellar/qt/4.7.3/lib/QtWebKit.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtWebKit -I/usr/local/Cellar/qt/4.7.3/include -I. -F/usr/local/Cellar/qt/4.7.3/lib -o moc_Page.o moc_Page.cpp
/usr/local/Cellar/qt/4.7.3/bin/moc -DQT_NO_DEBUG -DQT_WEBKIT_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/local/Cellar/qt/4.7.3/mkspecs/macx-g++ -I. -I/usr/local/Cellar/qt/4.7.3/lib/QtCore.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtCore -I/usr/local/Cellar/qt/4.7.3/lib/QtNetwork.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtNetwork -I/usr/local/Cellar/qt/4.7.3/lib/QtGui.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtGui -I/usr/local/Cellar/qt/4.7.3/lib/QtWebKit.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtWebKit -I/usr/local/Cellar/qt/4.7.3/include -I. -F/usr/local/Cellar/qt/4.7.3/lib -D__APPLE__ -D__GNUC__ HeadlessSpecRunner/Runner.h -o moc_Runner.cpp
g++ -c -pipe -O2 -Wall -W -DQT_NO_DEBUG -DQT_WEBKIT_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/local/Cellar/qt/4.7.3/mkspecs/macx-g++ -I. -I/usr/local/Cellar/qt/4.7.3/lib/QtCore.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtCore -I/usr/local/Cellar/qt/4.7.3/lib/QtNetwork.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtNetwork -I/usr/local/Cellar/qt/4.7.3/lib/QtGui.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtGui -I/usr/local/Cellar/qt/4.7.3/lib/QtWebKit.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.7.3/include/QtWebKit -I/usr/local/Cellar/qt/4.7.3/include -I. -F/usr/local/Cellar/qt/4.7.3/lib -o moc_Runner.o moc_Runner.cpp
g++ -headerpad_max_install_names -o jasmine-webkit-specrunner Page.o Runner.o specrunner.o moc_Page.o moc_Runner.o   -F/usr/local/Cellar/qt/4.7.3/lib -L/usr/local/Cellar/qt/4.7.3/lib -framework QtWebKit -framework QtGui -L/usr/X11/lib -L/usr/local/Cellar/qt/4.7.3/lib -F/usr/local/Cellar/qt/4.7.3/lib -framework QtCore -framework QtNetwork 


Gem files will remain installed in /Users/Thibaud/.rvm/gems/ruby-1.9.2-p290/gems/jasmine-headless-webkit-0.6.3 for inspection.
Results logged to /Users/Thibaud/.rvm/gems/ruby-1.9.2-p290/gems/jasmine-headless-webkit-0.6.3/ext/jasmine-webkit-specrunner/gem_make.out
```

Any idea? I have tried with the master branch but I got the same issue.
Thanks for your help :)

**Comment by johnbintz** [comment]
It seems to be hitting the linker...very odd. Is there a `jasmine-webkit-specrunner` executable in the `ext/jasmine-webkit-specrunner` directory? Can you run it directly? What if you run the `extconf.rb` file directly?

**Comment by scottdavis** [comment]
i got it running on my macbook using the download version of QT and lion

**Comment by johnbintz** [comment]
...I should add that any testing of compilation on Homebrew is being done by others, since I use MacPorts (LIKE A BOSS).

**Comment by thibaudgg** [comment]
Yes, there's a `jasmine-webkit-specrunner` executable in the `/Users/Thibaud/.rvm/gems/ruby-1.9.2-p290/gems/jasmine-headless-webkit-0.6.3/ext/jasmine-webkit-specrunner` directory. The output is:

```
→ ./jasmine-webkit-specrunner 
Run Jasmine's SpecRunner headlessly

  specrunner [-c] [-r <report file>] specrunner.html ...
```

When I execute `extconf.rb` I got `Permission denied` that could be the problem?

@jetviper21 not sure if I want to try the downloadable of QT version right now. I love too much Homebrew :) will see..

**Comment by johnbintz** [comment]
@thibaudgg, try `ruby extconf.rb` and, after running, check the exit status with `echo $?`. Something, somewhere is misbehaving...

**Comment by thibaudgg** [comment]
`ruby extconf.rb` => `make: Nothing to be done for 'first'.`
`echo $?` => `0`

**Comment by johnbintz** [comment]
`gem uninstall` your current build and try `rake install`ing the latest HEAD, except this time, modify the `extconf.rb` file to specifically `exit 0` before doing so. Let's see if that makes a difference...

**Comment by johnbintz** [comment]
Also, try this...go into the dir with the `Makefile` and, with the generated `Makefile` in there and having already built everything, run: `make && echo $? && make -v`. I wonder if it's an issue with Ruby's use of `make` (though I could be 100% wrong :) ).

**Comment by johnbintz** [comment]
.,..or rather, my poor understanding of `extconf.rb`...

**Comment by thibaudgg** [comment]
Ok thanks, I'll try this next tuesday (when I'll be behind my MacBook).

**Comment by thibaudgg** [comment]
Ok so...
- Modifying the `extconf.rb` do not made any differences
- not sure for the `Makefile` with this at `/Users/Thibaud/.rvm/gems/ruby-1.9.2-p290/gems/jasmine-headless-webkit-0.6.3/ext/jasmine-webkit-specrunner/Makefile` with `make && echo $? && make -v` I got:

```→ make && echo $? && make -v
make: Nothing to be done for`first'.
0
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.0

```

Also note, that I have tried to install it on 1.8.7-p352 (with RVM) and it was working great! Weird right? Maybe there's a problem with 1.9.2?
```

**Comment by thibaudgg** [comment]
It's also not working with `ruby-1.9.3-preview1` :(

**Comment by johnbintz** [comment]
Did you install Qt from source using Homebrew, or did you do the binary install? Another user let me know he had to install from source for JHW.

**Comment by thibaudgg** [comment]
I have tried with Homebrew & Binary install. Can you tell me how install it from JHW?

**Comment by johnbintz** [comment]
I was told `brew install qt --build-from-source` was necessary. Sadly I can't easily test it myself, as no Lion machine handy.

**Comment by thibaudgg** [comment]
After 3-4 hours of qt compiling `brew install qt --build-from-source` I still got the same issue... too bad :(

**Comment by johnbintz** [comment]
![fffffffuuuuuuuuuuuu](http://profile.ak.fbcdn.net/hprofile-ak-snc4/71176_193726421090_8349741_n.jpg)

...OK, what version of Rubygems do you have?

**Comment by johnbintz** [comment]
Also, I'm sitting on the #guard channel if you want to work on this real-time. We can paste in the solution here once we get it (and we _will_ get it!)

**Comment by thibaudgg** [comment]
Rubygems 1.8.10 (the last one). I'll be back behind my MacBook next thursday. I'll be on #guard channel. Thanks for your help!

**Comment by johnbintz** [comment]
A quick digging into the Rubygems source shows me this: https://github.com/rubygems/rubygems/blob/v1.8.10/lib/rubygems/installer.rb#L536

Whatever exception is happening is being eaten. Want to hack your copy of Rubygems to actually get the exception and print it to the user? That way we can actually get an idea of what's happening.

**Comment by thibaudgg** [comment]
Yep, I'll try that this Thursday :) Thanks

**Comment by changa** [comment]
I've experienced the same bug, and found the cause : the original exception swallowed by the rubygems installer is a `<ArgumentError: invalid byte sequence in UTF-8` when processing the Makefile generated by qt/make.

Apparently, the Makefile is generated using ISO-8859-1 charset by default. In most cases, that won't be a problem when rubygems parses it as UTF-8, but when you're on a French system like thibaudgg and myself and the current month is _août_ (august), the resulting comment : `# Generated by qmake (2.01a) (Qt 4.7.3) on: mer. août 31 12:50:01 2011` (ISO 8859-1 encoded) will make rubygems barf.

TL;DR : the short fix is `LC_ALL gem install jasmine-headless-webkit`

**Comment by johnbintz** [comment]
@changa, awesome detective work! Where is that comment located in the file? Can you gist it somewhere? I might as well strip that comment from the Makefile to simplify the installation procedure for next August, December, or February. :)

Also, booo @ Rubygems for swallowing that error.

**Comment by changa** [comment]
The comment is located just in the top comment block generated by qmake:
    #############################################################################
    # Makefile for building: jasmine-webkit-specrunner
    # Generated by qmake (2.01a) (Qt 4.7.3) on: mer. août 31 12:50:01 2011
    # Project:  specrunner.pro
    # Template: app
    # Command: /usr/bin/qmake -spec /usr/share/qt4/mkspecs/linux-g++ -o Makefile specrunner.pro
    #############################################################################

I guess having the generated extconf.rb run
    case RUBY_PLATFORM
    when /linux/
        system %{LC_ALL=C qmake -spec linux-g++}
    else
        system %{LC_ALL=C qmake -spec macx-g++}
    end

would prevent the problem. (I haven't had the time to test it yet, though).

**Comment by thibaudgg** [comment]
Awesome, can't wait to try that tomorrow :)

**Comment by changa** [comment]
You mean when it will be back to normal anyway because the month will be `septembre` ? :-)

**Comment by johnbintz** [comment]
@changa, you beat me to saying that to @thibaudgg. ;) Out of curiosity, can you provide the backtrace to that Rubygems error? I'd like to see exactly what's triggering that read of the Makefile.

**Comment by changa** [comment]
OK, @johnbintz, I've made a quick and dirty gist : https://gist.github.com/1183434.
The error is actually triggered [here](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/ext/builder.rb#L20), when Gem::Ext::Builder.make tries to apply a gsub on the Makefile.

**Comment by johnbintz** [comment]
OK I'm starting to understand how all this works, and I think I can even optimize how I'm doing the build on Gem install. We may not need to worry about the locale thing, I'll just need to work around some Rubygems and (older versions of) Bundler limitations.

**Comment by thibaudgg** [comment]
September 1st, I could be able to install it without any problem. Happy!. I have changed my formats settings ('Language & Text preference pane) to "United States (computer)". My whole system was already in english but the formats settings was on "Swiss (French)".
Thanks again for your precious help.

**Comment by johnbintz** [comment]
Dunno how we can best test this before December, but I did commit bd8fa3e which, after the compilation with the QMake-generated file is done, pops in a dummy Makefile just for Rubygems. Now that I understand how the process works, this should solve the problem once and for all.

**Comment by thibaudgg** [comment]
Great, thanks!

**Comment by changa** [comment]
Nice! Thank you!
